oi/FuncGen.cpp

```diff
@@ -391,3 +430,217 @@ void FuncGen::DeclareGetContainer(std::string& testCode) {
+            if constexpr(std::is_pointer_v<T>) {
+                return std::type_identity<StaticTypes::Pair<DB,
+                  StaticTypes::VarInt<DB>,
+                  StaticTypes::Sum<DB,
+                    StaticTypes::Unit<DB>,
+                    typename TypeHandler<DB, std::remove_pointer_t<T>>::type
+                >>>();
+            } else {
+                return std::type_identity<StaticTypes::Unit<DB>>();
+            }
+        }
+      public:
+        using type = typename decltype(choose_type())::type;
+        static StaticTypes::Unit<DB> getSizeType(
+          const T& t,
+          typename TypeHandler<DB, T>::type returnArg) {
+            if constexpr(std::is_pointer_v<T>) {
+              JLOG("ptr val @");
+              JLOGPTR(t);
+              auto r0 = returnArg.write((uintptr_t)t);
+              if (t && pointers.add((uintptr_t)t)) {
+                return r0.template delegate<1>([&t](auto ret) {
+                  if constexpr (!std::is_void<std::remove_pointer_t<T>>::value) {
+                    return TypeHandler<DB, std::remove_pointer_t<T>>::getSizeType(*t, ret);
+                  } else {
+                    return ret;
+                  }
+                });
+              } else {
+                return r0.template delegate<0>(std::identity());
+              }
+            } else {
+              return returnArg;
+            }
+          }
+    };
+    template <typename DB>
+    class TypeHandler<DB, void> {
+      public:
+        using type = StaticTypes::Unit<DB>;
+    };
+  )";
+
+  testCode.append(handlers);
+}
+
+void FuncGen::DefineStaticTypes(std::string& testCode) {
+  constexpr std::string_view unitType = R"(
+    template <typename DataBuffer>
+    class Unit {
+      public:
+        Unit(DataBuffer db) : _buf(db) {}
+
+        size_t offset() {
+          return _buf.offset();
+        }
+
+        template <typename T>
+        T cast() {
+          return T(_buf);
+        }
+
+        Unit<DataBuffer>
+        delegate(std::function<Unit<DataBuffer>(Unit<DataBuffer>)> cb) {
+          return cb(*this);
+        }
+
+      private:
+        DataBuffer _buf;
+    };
+  )";
+
+  constexpr std::string_view varintType = R"(
+    template <typename DataBuffer>
+    class VarInt {
+      public:
+        VarInt(DataBuffer db) : _buf(db) {}
+
+        Unit<DataBuffer> write(uint64_t val) {
+          while (val >= 128) {
+            _buf.write_byte(0x80 | (val & 0x7f));
+            val >>= 7;
+          }
+          _buf.write_byte(uint8_t(val));
+          return Unit<DataBuffer>(_buf);
+        }
+
+      private:
+        DataBuffer _buf;
+    };
+  )";
+
+  constexpr std::string_view pairType = R"(
+    template <typename DataBuffer, typename T1, typename T2>
+    class Pair {
+      public:
+        Pair(DataBuffer db) : _buf(db) {}
+        template <class U>
+        T2 write(U val) {
+          Unit<DataBuffer> second = T1(_buf).write(val);
```

Comment:
Is T1.write() guaranteed to always return a Unit?

Is using a Pair like this explicitly blocked in some way? Pseudo-code:
`Pair<Pair<VarInt, VarInt>, VarInt>`

Reply:
That type is valid, but calling `.write(...)` on it is blocked at compile time, because it won't return a `Unit`. The expected way to nest pairs is on the right hand side, then the `.write()` drops the top level pair and you're left with a new pair. If you did have a `Pair<Pair<...>, X>` you could instead call `.delegate()` with a callback to complete the left hand side and leave you with the right.